src/core/pointcloud/qgspointcloudrgbrenderer.cpp

```diff
@@ -0,0 +1,247 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgspointcloudrgbrenderer.h"
+#include "qgspointcloudblock.h"
+
+QgsPointCloudRgbRenderer::QgsPointCloudRgbRenderer()
+{
+
+}
+
+QString QgsPointCloudRgbRenderer::type() const
+{
+  return QStringLiteral( "rgb" );
+}
+
+QgsPointCloudRenderer *QgsPointCloudRgbRenderer::clone() const
+{
+  std::unique_ptr< QgsPointCloudRgbRenderer > res = qgis::make_unique< QgsPointCloudRgbRenderer >();
+  res->mPenWidth = penWidth();
+  res->mRedAttribute = mRedAttribute;
+  res->mGreenAttribute = mGreenAttribute;
+  res->mBlueAttribute = mBlueAttribute;
+  return res.release();
+}
+
+void QgsPointCloudRgbRenderer::renderBlock( const QgsPointCloudBlock *block, QgsPointCloudRenderContext &context )
+{
+  const QgsMapToPixel mapToPixel = context.renderContext().mapToPixel();
+
+  QgsRectangle mapExtent = context.renderContext().mapExtent();
+
+  QPen pen;
+  pen.setWidth( mPainterPenWidth );
+  pen.setCapStyle( Qt::FlatCap );
+  //pen.setJoinStyle( Qt::MiterJoin );
+
+  const char *ptr = block->data();
+  int count = block->pointCount();
+  const QgsPointCloudAttributeCollection request = block->attributes();
+
+  const std::size_t recordSize = request.pointRecordSize();
+  int redOffset = 0;
+  const QgsPointCloudAttribute *attribute = request.find( mRedAttribute, redOffset );
+  if ( !attribute )
+    return;
+  const QgsPointCloudAttribute::DataType redType = attribute->type();
+
+  int greenOffset = 0;
+  attribute = request.find( mGreenAttribute, greenOffset );
+  if ( !attribute )
+    return;
+  const QgsPointCloudAttribute::DataType greenType = attribute->type();
+
+  int blueOffset = 0;
+  attribute = request.find( mBlueAttribute, blueOffset );
+  if ( !attribute )
+    return;
+  const QgsPointCloudAttribute::DataType blueType = attribute->type();
+
+  int rendered = 0;
+  double x = 0;
+  double y = 0;
+  for ( int i = 0; i < count; ++i )
+  {
+    pointXY( context, ptr, i, x, y );
+
+    if ( mapExtent.contains( QgsPointXY( x, y ) ) )
+    {
+      int red = 0;
+      switch ( redType )
+      {
+        case QgsPointCloudAttribute::Char:
+          continue;
+
+        case QgsPointCloudAttribute::Int32:
+          red = *( qint32 * )( ptr + i * recordSize + redOffset );
+          break;
+
+        case QgsPointCloudAttribute::Short:
+          red = *( short * )( ptr + i * recordSize + redOffset );
+          break;
+
+        case QgsPointCloudAttribute::Float:
+          red = *( float * )( ptr + i * recordSize + redOffset );
+          break;
+
+        case QgsPointCloudAttribute::Double:
+          red = *( double * )( ptr + i * recordSize + redOffset );
+          break;
```

Comment:
seeing this one in various places and probably more to come... wondering if we continue this way or some template function/helper function somewhere to have reusable code?

Reply:
I agree... I was going to try adding a few more renderers and see how much more duplicate code we end up with, and then try to remove it 😂